network/crl.go

```diff
@@ -0,0 +1,380 @@
+// (as opposed to stub ones used in tests)
+type DefaultCRLClient struct{}
+
+// Fetch fetches CRL from passed URI (can be either http:// or file://)
+func (c DefaultCRLClient) Fetch(uri string) ([]byte, error) {
+	if strings.HasPrefix(uri, "http://") {
+		httpRequest, err := http.NewRequest(http.MethodGet, uri, nil)
+		if err != nil {
+			return nil, err
+		}
+		crlURL, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+		httpRequest.Header.Add("Accept", "application/pkix-crl, application/pem-certificate-chain")
+		httpRequest.Header.Add("host", crlURL.Host)
+		httpClient := &http.Client{}
+		httpResponse, err := httpClient.Do(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		defer httpResponse.Body.Close()
+		content, err := ioutil.ReadAll(httpResponse.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if httpResponse.StatusCode != 200 {
+			return nil, fmt.Errorf("Server returned non-OK status: %s", httpResponse.Status)
+		}
+
+		return content, nil
+	} else if strings.HasPrefix(uri, "file://") {
+		content, err := ioutil.ReadFile(uri[7:])
+		if err != nil {
+			return nil, err
+		}
+
+		return content, nil
+	}
+
+	return nil, fmt.Errorf("Cannot fetch CRL from '%s', unsupported protocol", uri)
+}
+
+// CRLCache is used to store fetched CRLs to avoid downloading the same URI more than once
+type CRLCache interface {
+	Get(key string) ([]byte, error)
+	Put(key string, value []byte) error
+	Remove(key string) error
+}
+
+// DefaultCRLCache is a default implementation of CRLCache, internally stores items in sync.Map
+type DefaultCRLCache struct {
+	cache sync.Map
+}
+
+// Get tries to get CRL from cache, returns error if failed
+func (c *DefaultCRLCache) Get(key string) ([]byte, error) {
+	log.Debugf("CRL: cache: loading '%s'", key)
+	value, ok := c.cache.Load(key)
+	if ok {
+		value, ok := value.([]byte)
+		if !ok {
+			// should never happen since Put() only inserts []byte
+			return nil, errors.New("Unexpected value of invalid type")
+		}
+		log.Debugf("CRL: cache: '%s' found", key)
+		return value, nil
+	}
+
+	log.Debugf("CRL: cache: '%s' not found", key)
+	return nil, errors.New("Key not found")
+}
+
+// Put stores CRL in cache
+func (c *DefaultCRLCache) Put(key string, value []byte) error {
+	c.cache.Store(key, value)
+	log.Debugf("CRL: cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *DefaultCRLCache) Remove(key string) error {
+	c.cache.Delete(key)
+	log.Debugf("CRL: cache: removed '%s'", key)
+	return nil
+}
+
+// ParsedCRLCache is similar to CRLCache, but works on top of it and stores parsed and verified CRLs
+type ParsedCRLCache interface {
+	Get(key string) (*pkix.CertificateList, error)
+	Put(key string, value *pkix.CertificateList) error
+	Remove(key string) error
+}
+
+// PRLRUParsedCRLCache is an implementation of ParsedCRLCache that uses LRU cache inside
+type LRUParsedCRLCache struct {
+	cache lru.Cache
```

Comment:
why not to re-use - https://github.com/cossacklabs/acra/blob/master/keystore/lru/cache.go this our cache implementation, which has same struct?) and only make method wrappers which parse bytes to final output structs

Reply:
Didn't reuse one from `keystore`  since:
\- key/value are `string`/`[]byte` (like in `func (cache *Cache) Add(keyID string, keyValue []byte)`) while we need `pkix.CertificateList` (or even some custom struct, to contain CRL and fetch time)
\- that's a small piece of code, it cannot be properly reused and I don't wanna change it just to make it fit into both keystore needs and CRL cache needs, potentially breaking something else